File: windows/tests/FlowPiano.Windows.Tests/LayoutVisibilityTests.cs
using FlowPiano.Windows.Core;
using Xunit;

namespace FlowPiano.Windows.Tests;

public sealed class LayoutVisibilityTests
{
    [Fact]
    public void StudioOnlyLayersAreExcludedFromPublicOutput()
    {
        var publicLayers = LayoutEngine.VisibleLayers(LayoutConfiguration.Default, RenderTarget.PublicOutput);

        Assert.DoesNotContain(publicLayers, layer => layer.Kind == LayerKind.MusicStaff);
        Assert.DoesNotContain(publicLayers, layer => layer.Kind == LayerKind.AudioMeters);
        Assert.DoesNotContain(publicLayers, layer => layer.Kind == LayerKind.MidiEventLog);
        Assert.DoesNotContain(publicLayers, layer => layer.Kind == LayerKind.LatencyIndicator);
        Assert.DoesNotContain(publicLayers, layer => layer.Kind == LayerKind.Diagnostics);
    }

    [Fact]
    public void ValidationDetectsPublicLeak()
    {
        var unsafeConfiguration = new LayoutConfiguration(
            LayoutConfiguration.Default.Layers.Select(layer =>
                layer.Kind == LayerKind.Diagnostics
                    ? layer with { Visibility = layer.Visibility with { PublicVisible = true } }
                    : layer).ToArray());

        Assert.Contains(LayerKind.Diagnostics, LayoutEngine.ValidatePublicOutput(unsafeConfiguration));
        Assert.Empty(LayoutEngine.ValidatePublicOutput(LayoutEngine.SanitizedForPublicOutput(unsafeConfiguration)));
    }
}

File: Tests/Unit/LayoutVisibilityTests.swift
import XCTest
@testable import LayoutEngine

final class LayoutVisibilityTests: XCTestCase {
    func testStudioOnlyLayersAreExcludedFromPublicOutput() {
        let publicLayers = LayoutEngine.visibleLayers(in: .default, for: .publicOutput)
        XCTAssertFalse(publicLayers.contains(where: { $0.kind == .musicStaff }))
        XCTAssertFalse(publicLayers.contains(where: { $0.kind == .audioMeters }))
        XCTAssertFalse(publicLayers.contains(where: { $0.kind == .midiEventLog }))
        XCTAssertFalse(publicLayers.contains(where: { $0.kind == .latencyIndicator }))
        XCTAssertFalse(publicLayers.contains(where: { $0.kind == .diagnostics }))
    }

    func testStudioMonitorContainsLocalOnlyLayers() {
        let studioLayers = LayoutEngine.visibleLayers(in: .default, for: .studioMonitor)
        XCTAssertTrue(studioLayers.contains(where: { $0.kind == .musicStaff }))
        XCTAssertTrue(studioLayers.contains(where: { $0.kind == .audioMeters }))
        XCTAssertTrue(studioLayers.contains(where: { $0.kind == .midiEventLog }))
        XCTAssertTrue(studioLayers.contains(where: { $0.kind == .latencyIndicator }))
        XCTAssertTrue(studioLayers.contains(where: { $0.kind == .diagnostics }))
    }

    func testValidationDetectsPublicLeak() {
        let unsafeConfiguration = LayoutConfiguration(
            layers: LayoutConfiguration.default.layers.map { layer in
                guard layer.kind == .diagnostics else { return layer }
                var updated = layer
                updated.visibility.publicVisible = true
                return updated
            }
        )

        XCTAssertEqual(LayoutEngine.validatePublicOutput(in: unsafeConfiguration), [.diagnostics])
        XCTAssertEqual(LayoutEngine.validatePublicOutput(in: LayoutEngine.sanitizedForPublicOutput(unsafeConfiguration)), [])
    }
}
